_includes/langselect.html

```diff
@@ -1,17 +1,39 @@
-<div id="langselect">
-  <div lang="en" id="suglang">This site might also be available in your native language. Select your preferred language below:</div>
-  <ul>
-    {% for lng in site.languages %}
-      {% if lng == site.active_lang %}
-        <li class="selected" data-slang="{{ lng }}">{{ lng }}</li>
-      {% else %}
-        {% if lng == site.default_lang %}
-          {% capture langlink %} {{ page.url }}{% endcapture %}
-        {% else %}
-          {% capture langlink %}/{{ lng }}{{ page.url }}{% endcapture %}
-        {% endif %}
-        <li><a href="{{ langlink }}" data-slang="{{ lng }}">{{ lng }}</a></li>
-      {% endif %}
-    {% endfor %}
-  </ul>
+<div id="langselect-container">
+  <div id="langselect">
+    <!-- Language Dropdown -->
+    <form action="" method="get" id="language-form">
+      <div class="language-dropdown">
+        <select id="language-select" name="language-select" onchange="location = this.value;">
+          {% assign sorted_languages = site.languages | sort %}
+          {% for lng in sorted_languages %}
+            {% if lng == site.default_lang %}
+              {% capture langlink %} {{ page.url }}{% endcapture %}
+              {% capture lang_rel %}canonical{% endcapture %}
+            {% else %}
+              {% capture langlink %}{{ site.baseurl }}/{{ lng }}{{ page.url }}{% endcapture %}
+              {% capture lang_rel %}alternate{% endcapture %}
+            {% endif %}
+            <option value="{{ langlink }}" {% if lng == site.active_lang %}selected{% endif %}>
+              {{ site.language_names[lng] }}
+            </option>
+          {% endfor %}
+        </select>
+      </div>
+    </form>
+    <!-- Plain list of language links for non-JS users -->
+    <noscript>
+      <ul class="language-list">
+        {% for lng in sorted_languages %}
+          {% if lng == site.default_lang %}
+            {% capture langlink %} {{ page.url }}{% endcapture %}
+            {% capture lang_rel %}canonical{% endcapture %}
+          {% else %}
+            {% capture langlink %}{{ site.baseurl }}/{{ lng }}{{ page.url }}{% endcapture %}
+            {% capture lang_rel %}alternate{% endcapture %}
+          {% endif %}
+          <li><a href="{{ langlink }}">{{ site.language_names[lng] }}</a></li>
```

Comment:
@gilgongo: not sure if something needs to be added here for the canonical links. And also a bit further up, line 16.